seqr/views/apis/anvil_workspace_api.py

```diff
@@ -0,0 +1,118 @@
+    """
+    Redirect to the loading data from workspace page or redirect to the project if the project exists.
+
+    Validate the workspace and project before loading data.
+
+    :param request: Django request object
+    :param namespace: The namespace (or the billing account) of the workspace
+    :param name: The name of the workspace. It also be used as the project name
+    :return Redirect to a page depending on if the workspace permissions or project exists
+    Error page if the workspace doesn't exist or the user doesn't have a WRITER permission.
+    Redirect to '/create_project_from_workspace/:workspace_namespace/:workspace_name' if
+    project doesn't exist. Redirect to '/project/:projectGuid'
+
+    """
+    if _workspace_can_edit(request.user, namespace, name):
+        project = Project.objects.filter(name=name)
+        if project:
+            return redirect('/project/{}'.format(project.guid))
+        else:
+            return redirect('/create_project_from_workspace/{}/{}'.format(namespace, name))
+    return {"Error: Workspace {}/{} doesn't exist or misses required permissions (e.g. WRITER, OWNER, or PROJECT_OWNER)"
+                .format(namespace, name)}
+
+
+@login_required(login_url=API_LOGIN_REQUIRED_URL)
+def create_project_from_workspace(request, namespace, name):
+    """
+    Create a project when a cooperator requests to load data from an AnVIL workspace.
+
+    :param request: Django request object
+    :param namespace: The namespace (or the billing account) of the workspace
+    :param name: The name of the workspace. It also be used as the project name
+    :return the projectsByGuid with the new project json
+    """
+    # Validate that the current user has logged in through google and has one of the valid can_edit levels of
+    #  access on the specified workspace
+    project = Project.objects.filter(name = name)
+    if project:
+        error = 'Project {} exists.'.format(name)
+        return create_json_response({'error': error}, status=400, reason=error)
+
+    if not _workspace_can_edit(request.user, namespace, name):
+        return create_json_response({
+            'projectsByGuid': {},
+        })
+
+    # Validate all the user inputs from the post body
+    request_json = json.loads(request.body)
+
+    missing_fields = [field for field in ['genomeVersion', 'uploadedFileId'] if not request_json.get(field)]
+    if missing_fields:
+        error = 'Field(s) "{}" are required'.format(', '.join(missing_fields))
+        return create_json_response({'error': error}, status=400, reason=error)
+
+    if not request_json.get('agreeSeqrAccess'):
+        error = 'Must agree to grant seqr access to the data in the associated workspace.'
+        return create_json_response({'error': error}, status = 400, reason = error)
+
+    # Add the seqr service account to the corresponding AnVIL workspace
+    if add_service_account(request.user, namespace, name):
```

Comment:
Using if-else block in validation like this is really confusing to read, as I have to scroll down 30 lines to understand what happens if the check fails. Just like we do everywhere else is seqr, including above in this function, you want to check and immediately return an error if validation fails, not put the success code in an if block

```
success = add_service_account(request.user, namespace, name)
if not success:
    error = 'Failed to grant seqr service account access to the workspace'
    return create_json_response({'error': error}, status=400, reason=error)
# continue with the project creation
...
```
Alternatively, you might get a more detailed error for *why* adding the account failed if you throw descriptive exceptions in that function. So for example, instead of returning False you could  do 
```
raise  TerraAPIException('Failed to grant seqr service account access to the workspace', 400)
```
Since our middleware will handle any `TerraAPIException` for us, you wouldn't even need to do anything special here you could just do

```
add_service_account(request.user, namespace, name)
# continue with the project creation
...
```

Reply:
okay